### ISSUE TITLE
Developer's "crazy ideas" and TODO checklist for the gory details.

### ISSUE THREAD

**Issue by GerHobbelt** [jimmejardine/qiqqa-open-source#296]
# This particular issue is *For Developers Only*

---

**Please file a fresh, new issue if you see something you want to request as a feature or report a bug on or simply talk about.**

**Copy&paste the bit of text that's relevant, if you want.**

---

The notorious "*Mother Of All PRs*" for Qiqqa. (Folks who've worked with me before, will know the feeling. 😉🤯 ) 

## Why?

Because I don't want to swamp the issue tracker with the stuff I note, think about, or otherwise need to remind myself about at some later point in time, where my brain *very probably* has already given up attempting to track and manage.

a.k.a. "*Notes To Self*"

> I've been pondering dropping this stuff in the [kanban projects](https://github.com/jimmejardine/qiqqa-open-source/projects) but that hasn't worked out before, because this is about a **lot of details** that *do matter*, but are *clutter* for everyone trying to get a handle on the overall state of affairs (like *me*), so after a long time considering and trying other means, failures all, the current idea is to bundle all these devils in a single issue, with perhaps some check boxes, and then using the github EDIT feature, rather than COMMENTing each time: once an item in this list is done, it can be DELETED as far as I'm concerned. The git commit log serves well for keeping track of what happened and what was done; it's the 'gotta do / check this' buggers that need a place to go and I want to keep it all in a single website, i.e. *github*.

# Observed Crappiness?

- [ ] BibTeX editor. OMG. 🤯 I know it's hacky (hey, *I* did that back in '19, **I know**), but the parser is... 😒 and the (re)formatter... and the editor modes are 🤡  -- find time and effort to do a proper editor one day, please. Doesn't need to be smart, just **flexible**. **Tolerant**, say.
- [ ] xxx
- [ ] xxx

# *Crazy Ideas* To Try?

- [ ] [ReoGrid](https://github.com/unvell/ReoGrid) spreadsheet bulk editing of metadata 
  + [ ] free export to Excel that way?
  + [ ] how about letting user edit it in another tool, then re-import the spreadsheet and bulk-update the metadata that way?
- [ ] I *loath* WPF (OK, *we already knew that*, tell me something *new*, will ya?); ditch the crappy table in the Sync Manager dialog and replace it with a [ReoGrid](https://github.com/unvell/ReoGrid) instance, i.e. make it act like a spreadsheet too. [ReoGrid](https://github.com/unvell/ReoGrid) says they can do **readonly** on cells. Yay! 🥳 
- [ ] use a Nuke-like node editor interface for the OCR process. Find a way to have it communicate with mupdf+leptonica+tesseract. 
  - 🚫  As different PDFs require different processing, should we store the graph per PDF?
  - [x] Better to link PDFs to a "OCR template" of some kind: that way we get the user automatically keep them "grouped" that way.
  - [ ] Ditto for text extraction! 
    + **Remember your own bloody OnSemi (ex-Fairchild) datasheets with obnoxious cover sheets!**
    + Remember those Taiwanese Uni PDF papers you have, which have their own flavour of more-or-less useful cover sheets to strip off / *reduce*.
    + Oh! Oh! The cover sheets of that scientific PDF aggregation site you got them from. Suppressed the name, but you know which one I mean! Almost white sheet, images of the authors and a bit of crap. Making it hard to recognize the papers as duplicates when you get them from elsewhere.
- [ ] Check NodeJS interaction speed: JSON chat + background 'webserver'? 
- [ ] Big Binary Chunks coming from background processes (node, [mupdf](https://github.com/ArtifexSoftware/mupdf), tesseract, [solr](https://lucene.apache.org/solr/), whatever): Named pipes for bigger loads or is that overkill and do we push everything through a socket anyway? (copy bin blocks vs memory mapped I/O for large data chunks)
- [ ] [tesseract](https://github.com/tesseract-ocr/tesseract/issues/2961) is [**integrated with mupdf by Artifex**](https://github.com/ArtifexSoftware/mupdf/search?q=tesseract&type=commits) 👍 👍 👍 🥳  -- now open that bugger up and permit scriptable/configurable image processing **inside** [mupdf-as-you-have-it](https://github.com/GerHobbelt/mupdf), so we can optimize OCR. Drive this with the [Nuke](https://www.foundry.com/products/nuke/features#link112)-style node editor mentioned above.
  + I mention Nuke, because that's what my brain comes up with for this Rorschach Blot, but it's more like Autodesk 3dsmax's Material Editor and their ilk: intermediate or fundamental as visual thumbnail in each node to help track where happens what:
    - https://unity3d.com/profiles/unity3d/themes/unity/images/pages/shader-graph/ShaderGraphHologram.jpg
    - https://cdn.soft8soft.com/images/materials_maps_max.jpg
  + [ ]  consider adding [OpenCV lib](https://docs.opencv.org/master/d7/dbd/group__imgproc.html) to the mupdf tool, next to [leptonica](http://leptonica.org/) (which is already the *default* part of the pdf-to-tesseract path in there), so we have additional image filters / processing for the OCR workflow.
- [ ] add a 'webserver' mode to mupdf (**mutool + mudraw**: both for metadata extraction, text extraction, OCR processing (and consequent text extraction) and PDF page to image rendering) so we don't have the (costly?!) tool startup time for every page or bit of (meta)data we're requesting.
  - [ ] be prepared to kill the bugger or expect the bastard to lock up or crash due to **nasty PDF inputs** once every couple of documents: feed the entire evil library through it, and then everything else you can grab off the Net. 
  
    **Reminder**: the SynFusion libs b0rked out with a HUGE memory leakage just today, and that was only because Qiqqa was doing a bit of *annotations* extraction via that one today. We got rid of SORAX, but SynFusion is on its way out too. 💢 

  - [ ] see if I should revive [my old mongoose clone](https://github.com/GerHobbelt/mongoose-vanilla) for this, or grab another *light*, **embeddable** C/C++ web server that can do JSON and a whole lot more.
    - [ ] what's it called now? Or was it the original with MIT license that got renamed? Don't recall precisely. There was a license change back then with some brouhaha, but since Qiqqa and mupdf are (A)GPL, it doesn't matter no more, right?!
    - [ ] I'm not looking at nginx and its ilk: way to much overkill and I've working with them before: I want an **embeddable** webserver, which can run on `localhost` for as long as Qiqqa is *alive*: started and stopped by Qiqqa, preferably. 
- [ ] Analyze Qiqqa PDF page render behaviour (we're doing that now) and see what we can do to improve performance there by caching some sort of LIFO / timestamped cache, where we can *age off* old slots. No more 3 images per PDF static crap.
- [ ] **Please** wrack your brain for a **smart solution** for the many 'get it as we need them' situations in the code, where we SafeThreadPool the work, but **should** ideally should:
  - [ ] respond with a **fake answer** immediately, so the UI gets rendered pronto. (Think the grey boxes on page load you see happenin' with modern websites, which perform async content loading)
  - [ ] don't care about effin' WPF INotifyWhatWasIt crap and code bloat, "required" for UI updating. Come up with something leaner and preferably faster / as fast once the 'async' data finally arrives.
  - [ ] **extra bonus points** when you go through the code and make sure the final result is checked against the current state of affairs in the UI to ensure the data is not already outdated, because the user has moved on and another panel or PDF is currently in view, while the work is about a PDF that's already closed again.
  - [ ] ** extra extra bonus points** if you can come up with a solid schema where all this work is *abortable* without a lot of hassle. It happens in a zillion places and I'ld rather not clutter the whole codebase with [CancellationTokens processing](https://docs.microsoft.com/en-us/dotnet/standard/threading/cancellation-in-managed-threads), if I can get away with something leaner, preferably **much** leaner.
